File: python/flask/sample/views/merge_cades_express.py
"""

This sample performs a merge of CAdES signature using PKI Express.

"""
import os
import uuid

from flask import Blueprint
from flask import render_template
from flask import current_app

from pkiexpress import CadesSignatureEditor
from sample.storage_mock import create_app_data
from sample.utils import set_pki_defaults


blueprint = Blueprint(os.path.basename(__name__), __name__,
                      url_prefix='/merge-cades-express')


@blueprint.route('/detached/<data_file_id>/<file_id1>/<file_id2>')
def detached(data_file_id, file_id1, file_id2):
    """

    This function performs a merge of CAdES signature using PKI Express
    when both signatures doesn't have encapsulated content

    """
    # Get an instance of the CadesSignatureEditor class, responsible for
    # receiving the files and merge them.
    signature_editor = CadesSignatureEditor()

    # Set PKI default options (see utils.py).
    set_pki_defaults(signature_editor)
    
    # Guarantees that "app data" folder exists.
    create_app_data()

    # Generate output filename
    output_file = '%s.p7s' % (str(uuid.uuid4()))

    # Set the CMS data file
    signature_editor.set_data_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], data_file_id))

    # Add both signatures
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id1))
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id2))

    # Set path to output file
    signature_editor.output_file = os.path.join(os.path.join(current_app.config['APPDATA_FOLDER'], output_file))

    # Merge files
    signature_editor.merge()

    return render_template('merge_cades_express/index.html', output_file=output_file)


@blueprint.route('/attached/<file_id1>/<file_id2>')
def attached(file_id1, file_id2):
    """

    This function performs a merge of CAdES signature using PKI Express
    when both signatures have encapsulated content

    """
    # Get an instance of the CadesSignatureEditor class, responsible for
    # receiving the files and merge them.
    signature_editor = CadesSignatureEditor()

    # Set PKI default options (see utils.py).
    set_pki_defaults(signature_editor)
    
    # Guarantees that "app data" folder exists.
    create_app_data()

    # Generate output filename
    output_file = '%s.p7s' % (str(uuid.uuid4()))

    # Add both signatures
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id1))
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id2))

    # Set path to output file
    signature_editor.output_file = os.path.join(current_app.config['APPDATA_FOLDER'], output_file)
    
    # Merge files
    signature_editor.merge()
    return render_template('merge_cades_express/index.html', output_file=output_file)


@blueprint.route('/mixed/<data_file_id>/<file_id1>/<file_id2>')
def mixed(data_file_id, file_id1, file_id2):
    """

    This function performs a merge of CAdES signature using PKI Express
    when one signature has encpasulated content and the other doesn't

    """
    # Get an instance of the CadesSignatureEditor class, responsible for
    # receiving the files and merge them.
    signature_editor = CadesSignatureEditor()

    # Set PKI default options (see utils.py).
    set_pki_defaults(signature_editor)
    
    # Guarantees that "app data" folder exists.
    create_app_data()

    # Generate output filename
    output_file = '%s.p7s' % (str(uuid.uuid4()))

    # Set the CMS data file
    signature_editor.set_data_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], data_file_id))

    # Add both signatures
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id1))
    signature_editor.add_cms_file_from_path(os.path.join(current_app.config['APPDATA_FOLDER'], file_id2))

    # Set path to output file
    signature_editor.output_file = os.path.join(current_app.config['APPDATA_FOLDER'], output_file)

    # Merge files
    signature_editor.merge()

    return render_template('merge_cades_express/index.html', output_file=output_file)

File: nodejs/expressmvc/routes/merge-cades-express.js
const express = require('express');
const path = require('path');
const uuidv4 = require('uuid/v4');
const { CadesSignatureEditor } = require('pki-express');

const { Util } = require('../util');
const { StorageMock } = require('../storage-mock');

const router = express.Router();
const APP_ROOT = process.cwd();


/*
 * GET /merge-cades-express/detached
 *
 * This sample performs a merge of CAdES signature using PKI Express
 * when both signatures doesn't have encapsulated content
 *
 */
router.get('/detached', (req, res) => {
	const { fileId1 } = req.query;
	const { fileId2 } = req.query;
	const { dataFileId } = req.query;
	const signatureEditor = new CadesSignatureEditor();
	Util.setPkiDefaults(signatureEditor);

	const outputFile = `${uuidv4()}.ps7`;

	signatureEditor.setDataFileFromPathSync(StorageMock.getDataPath(dataFileId));
	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId1));
	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId2));
	signatureEditor.setOutputFilePath(path.join(APP_ROOT, 'app-data', outputFile));

	signatureEditor.merge()
		.then(() => {
			res.render('merge-cades-express', { outputFile: outputFile.replace('.', '_') });
		});
});

/*
 * GET /merge-cades-express/attached
 *
 * This sample performs a merge of CAdES signature using PKI Express
 * when both signatures has encapsulated content
 *
 */
router.get('/attached', (req, res) => {
	const { fileId1 } = req.query;
	const { fileId2 } = req.query;

	const signatureEditor = new CadesSignatureEditor();
	Util.setPkiDefaults(signatureEditor);

	const outputFile = `${uuidv4()}.ps7`;

	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId1));
	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId2));
	signatureEditor.setOutputFilePath(path.join(APP_ROOT, 'app-data', outputFile));
	signatureEditor.merge()
		.then(() => {
			res.render('merge-cades-express', { outputFile: outputFile.replace('.', '_') });
		});
});

/*
 * GET /merge-cades-express/mixed
 *
 * This sample performs a merge of CAdES signature using PKI Express
 * when one signature has encpasulated content and the other doesn't
 *
 */
router.get('/mixed', (req, res) => {
	const { fileId1 } = req.query;
	const { fileId2 } = req.query;
	const { dataFileId } = req.query;

	const signatureEditor = new CadesSignatureEditor();
	Util.setPkiDefaults(signatureEditor);

	const outputFile = `${uuidv4()}.ps7`;

	signatureEditor.setDataFileFromPathSync(StorageMock.getDataPath(dataFileId));
	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId1));
	signatureEditor.addCmsFileFromPathSync(StorageMock.getDataPath(fileId2));
	signatureEditor.setOutputFilePath(path.join(APP_ROOT, 'app-data', outputFile));
	signatureEditor.merge()
		.then(() => {
			res.render('merge-cades-express', { outputFile: outputFile.replace('.', '_') });
		});
});


module.exports = router;
